model/map/src/main/java/org/keycloak/models/map/events/MapEventStoreProvider.java

```diff
@@ -0,0 +1,142 @@
+import org.keycloak.models.map.storage.ModelCriteriaBuilder;
+import org.keycloak.models.map.storage.QueryParameters;
+import org.keycloak.models.map.storage.criteria.DefaultModelCriteria;
+
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import static org.keycloak.models.map.events.EventUtils.modelToEntity;
+
+public class MapEventStoreProvider implements EventStoreProvider {
+
+    private final KeycloakSession session;
+    private final MapKeycloakTransaction<MapLoginEventEntity, Event> loginEventsTX;
+    private final MapKeycloakTransaction<MapAdminEventEntity, AdminEvent> adminEventsTX;
+
+    public MapEventStoreProvider(KeycloakSession session, MapStorage<MapLoginEventEntity, Event> loginEventsStore, MapStorage<MapAdminEventEntity, AdminEvent> adminEventsStore) {
+        this.session = session;
+        this.loginEventsTX = loginEventsStore.createTransaction(session);
+        this.adminEventsTX = adminEventsStore.createTransaction(session);
+
+        session.getTransactionManager().enlistAfterCompletion(this.loginEventsTX);
+        session.getTransactionManager().enlistAfterCompletion(this.adminEventsTX);
+    }
+
+    /** LOGIN EVENTS **/
+    @Override
+    public void onEvent(Event event) {
+        String id = event.getId();
+        if (id != null && loginEventsTX.read(id) != null) {
+            throw new ModelDuplicateException("Event already exists: " + id);
+        }
+
+        loginEventsTX.create(modelToEntity(event));
+    }
+
+    @Override
+    public EventQuery createQuery() {
+        return new MapLoginEventQuery(((Function<QueryParameters<Event>, Stream<MapLoginEventEntity>>) loginEventsTX::read)
+                .andThen(s -> s.map(EventUtils::entityToModel)));
+    }
+
+    @Override
+    public void clear() {
+        loginEventsTX.delete(QueryParameters.withCriteria(DefaultModelCriteria.criteria()));
+    }
+
+    @Override
+    public void clear(String realmId) {
+        loginEventsTX.delete(QueryParameters.withCriteria(DefaultModelCriteria.<Event>criteria()
+                .compare(Event.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realmId)));
+    }
+
+    @Override
+    public void clear(String realmId, long olderThan) {
+        loginEventsTX.delete(QueryParameters.withCriteria(DefaultModelCriteria.<Event>criteria()
+                .compare(Event.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realmId)
+                .compare(Event.SearchableFields.TIME, ModelCriteriaBuilder.Operator.LT, olderThan)
+        ));
+    }
+
+    @Override
+    public void clearExpiredEvents() {
+        // TODO: Find another way of doing this
+        session.realms().getRealmsStream().forEach(realm -> {
+            if (realm.isEventsEnabled() && realm.getEventsExpiration() > 0) {
+                long olderThan = Time.currentTimeMillis() - realm.getEventsExpiration() * 1000;
+                clear(realm.getId(), olderThan);
+            }
+        });
```

Comment:
@hmlnarik I am not sure it is the correct thing to do for events. Maybe it makes sense to have expiration changeable for events that are already present in the database. For example, if there are too many events in the DB and an administrator decides to introduce expiration time (there was no expiration set before), I would say it is the correct thing to remove events that are older than the new expiration time rather than apply this only to new events and basically preserve the old forever. WDYT?   

btw. there is also a test for this use case: https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/events/EventStoreProviderTest.java#L211

Reply:
Imagine that there would be 10000 realms. `session.realms().getRealmsStream()` would load all of them.
We need to be smarter than loading all realms.
Where is the `clearExpiredEvents` called from?

Reply:
Done, now we are using expiration field for removing expired events.